src/middlewared/middlewared/alert/source/snapshot_count.py

```diff
@@ -43,11 +42,28 @@ async def check(self):
                 key=None,
             )
 
-        for dataset in sorted(datasets.keys()):
-            count = datasets[dataset]
-            if count > max_:
-                return Alert(
-                    SnapshotCountAlertClass,
-                    {"dataset": dataset, "count": count, "max": max_},
-                    key=None,
-                )
+    async def _check_smb(self) -> list[Alert]:
+        """Return an `Alert` for every dataset shared over smb whose number of snapshots exceeds the limit."""
+        max_ = await self.middleware.call("pool.snapshottask.max_count")
+        datasets = await self.middleware.call("zfs.snapshot.count")
+        smb_shares = await self.middleware.call("sharing.smb.query")
```

Comment:
We're creating a bunch of unnecessary variables here which is memory inefficient. You're also now making duplicate calls to `zfs.snapshot.count` unnecessarily. The caller should initialize that variable and pass it around to the other methods. We can simplify this entire method quite a bit. I'd do something like this:
```
async def _check_smb(self, datasets: dict) -> list[Alert]:
    max_ = await self.middleware.call("pool.snapshottask.max_count")
    to_alert = list()
    for i in await self.middleware.call('sharing.smb.query'):
        count = datasets.get(i['path'].removeprefix('/mt/'), -1)
        if count > max_:
            to_alert.append(...

Reply:
I do like this better.